flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,381 @@
+import org.apache.flink.runtime.clusterframework.ContaineredTaskManagerParameters;
+import org.apache.flink.runtime.clusterframework.TaskExecutorResourceSpec;
+import org.apache.flink.runtime.clusterframework.TaskExecutorResourceUtils;
+import org.apache.flink.runtime.clusterframework.types.ResourceID;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.entrypoint.ClusterInformation;
+import org.apache.flink.runtime.entrypoint.parser.CommandLineOptions;
+import org.apache.flink.runtime.heartbeat.HeartbeatServices;
+import org.apache.flink.runtime.highavailability.HighAvailabilityServices;
+import org.apache.flink.runtime.metrics.groups.ResourceManagerMetricGroup;
+import org.apache.flink.runtime.resourcemanager.JobLeaderIdService;
+import org.apache.flink.runtime.resourcemanager.ResourceManager;
+import org.apache.flink.runtime.resourcemanager.exceptions.ResourceManagerException;
+import org.apache.flink.runtime.resourcemanager.slotmanager.SlotManager;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.rpc.RpcService;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Kubernetes specific implementation of the {@link ResourceManager}.
+ */
+public class KubernetesResourceManager extends ResourceManager<KubernetesWorkerNode>
+	implements FlinkKubeClient.PodCallbackHandler {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesResourceManager.class);
+
+	/** The taskmanager pod name pattern is {clusterId}-{taskmanager}-{attemptId}-{podIndex}. */
+	private static final String TASK_MANAGER_POD_FORMAT = "%s-taskmanager-%d-%d";
+
+	private final ConcurrentMap<ResourceID, KubernetesWorkerNode> workerNodeMap;
+
+	private final int numberOfTaskSlots;
+
+	private final TaskExecutorResourceSpec taskExecutorResourceSpec;
+
+	private final int defaultTaskManagerMemoryMB;
+
+	private final double defaultCpus;
+
+	private final Collection<ResourceProfile> slotsPerWorker;
+
+	private final Configuration flinkConfig;
+
+	/** Flink configuration uploaded by client. */
+	private final Configuration flinkClientConfig;
+
+	/** When ResourceManager failover, the max attempt should recover. */
+	private final AtomicLong currentMaxAttemptId = new AtomicLong(0);
+
+	private final AtomicLong currentMaxPodId = new AtomicLong(0);
```

Comment:
Actually i did not see how this field was used in the codes. Could you give some descriptions for it especially for the difference with above `currentMaxAttemptId`.

Reply:
`
		final String podName = String.format(
			TASK_MANAGER_POD_FORMAT,
			clusterId,
			currentMaxAttemptId.get(),
			currentMaxPodId.addAndGet(1));
`
Current max pod index. When creating a new pod, it should increase one.